encode.go

```diff
@@ -30,30 +30,61 @@ import (
 	"go.yaml.in/yaml/v4/internal/libyaml"
 )
 
+// Sentinel values for newEncoder parameters.
+// These provide clarity at call sites, similar to http.NoBody.
+var (
+	noWriter           io.Writer                 = nil
+	noOptions          *options                  = nil
+	noVersionDirective *libyaml.VersionDirective = nil
+	noTagDirective     []libyaml.TagDirective    = nil
+)
+
 type encoder struct {
-	emitter  libyaml.Emitter
-	event    libyaml.Event
-	out      []byte
-	flow     bool
-	indent   int
-	doneInit bool
+	emitter               libyaml.Emitter
+	event                 libyaml.Event
+	out                   []byte
+	flow                  bool
+	indent                int
+	lineWidth             int
+	doneInit              bool
+	explicitStart         bool
+	explicitEnd           bool
+	flowSimpleCollections bool
 }
 
-func newEncoder() *encoder {
+// newEncoder creates a new YAML encoder with the given options.
+//
+// The writer parameter specifies the output destination for the encoder.
+// If writer is nil, the encoder will write to an internal buffer.
+func newEncoder(writer io.Writer, opts *options) *encoder {
 	e := &encoder{
 		emitter: libyaml.NewEmitter(),
 	}
-	e.emitter.SetOutputString(&e.out)
-	e.emitter.SetUnicode(true)
-	return e
-}
 
-func newEncoderWithWriter(w io.Writer) *encoder {
-	e := &encoder{
-		emitter: libyaml.NewEmitter(),
+	// Apply options if provided, otherwise use defaults
+	if opts != nil {
+		e.indent = opts.indent
+		e.lineWidth = opts.lineWidth
+		e.emitter.CompactSequenceIndent = opts.compactSeqIndent
+		e.emitter.SetWidth(opts.lineWidth)
```

Comment:
There is an inconsistency here - we only call `e.emitter.SetWidth(opts.lineWidth)` when options are passed. We should move this call after the `if`.

Reply:
I think it would be better to always use an options object, instead of having two separate code paths for the defaults (depending if an option is passed or not).

Reply:
Opened PR #224